blockprod/src/detail/tests.rs

```diff
@@ -1000,7 +1000,7 @@ mod produce_block {
             )
         };
 
-        let blocks_to_generate = rng.gen_range(100..=1000);
+        let blocks_to_generate = rng.gen_range(20..=100);
```

Comment:
If this is taking a long time due to difficulty issues, it's probably a good idea to limit the difficulty so that it doesn't take a long time, instead of setting these arbitrary numbers that have no inherent guarantees. In practice the generation should have a time bound (statistically speaking).

Reply:
I think we're already limiting difficulty by using `create_unittest_pos_config()` - it sets `target_limit` to `Uint256::MAX`.

Ah, I guess I can set a timeout for the test itself, but would we want to automatically treat that as a pass if it times out? I don't think that's a good idea because we may never see it fail then

Reply:
> I think we're already limiting difficulty by using `create_unittest_pos_config()` - it sets `target_limit` to `Uint256::MAX`.

That's probably the upper limit for the target (hence limit on the easiest possible difficulty). You need the opposite limitation. Please double-check.

> Ah, I guess I can set a timeout for the test itself, but would we want to automatically treat that as a pass if it times out? I don't think that's a good idea because we may never see it fail then

No. That's not what I meant. I meant limiting the difficulty by never allowing it to be too hard.

Reply:
Yep, you were right - the docs say `target_limit` is the lowest difficulty bound meaning the hardest, so I was interpreting this incorrectly.

If this is the case, then I'm thinking the following line in `consensus/src/pos/target.rs` in `calculate_new_target()` should be changed to `max()`:

```
    let new_target = std::cmp::min(new_target, pos_config.target_limit());
```

... maybe @azarovh can have a look at `calculate_new_target()` to see if that's correct.

Either way, I've made a local change which sets `difficulty_change_limit` to 0. This essentially fixes `new_target` difficulty during each calculation to the easiest PoS difficulty. So for PoS on my machine, I'm getting around 20 blocks per second, which gets us close to the 1 minute timeout for a unit test when we try and generate 1000 blocks.

I know it's arbitrary, think if I keep `let blocks_to_generate = rng.gen_range(20..=100);` and also fix `difficulty_change_limit`, then this particular unit test should never hit the timeout even in the worst case while still exercising the various consensus `NetUpgrades`.

Reply:
> If this is the case, then I'm thinking the following line in consensus/src/pos/target.rs in calculate_new_target() should be changed to max():

Changing this to `max` doesn't seem right because it will always be true. 
Keeping `difficulty_change_limit` at 0 is more reasonable fix. It actually says that for tests we don't want the difficulty to change at all.

I'll take a look if I can see other options to fix that for tests.

Reply:
At this point I'm not sure that there is a particular problem with this test or retargeting algo. If running with the same Seed changing `difficulty_change_limit` doesnt effect result at all.
Also I tried disabling PoS consensus upgrade or PoW. And the test is just slow anyway.

I have a profiler result if you are interested in looking deeper. 
![flamegraph](https://github.com/mintlayer/mintlayer-core/assets/105876856/0aba14d4-8c9a-4e46-ab1e-2882e0bf70d1)

Reply:
Thanks for that. So for each of the consensus types, I try to generate the blocks using the other types before I create the real one. However at the moment the JobManager takes a custom Id, but it's it's `None`, it will just use that. But before I can call `produce_block()` again, I have to wait for the JobManager to clean up.

I'm thinking that at the moment it's not cleaning up as fast as I thought... completely removing `Ignore` and `PoW`, then as @TheQuantumPhysicist said, `PoS` flies - 1000 `PoS` blocks in 10 seconds.

... so the difficulty was another red herring I think.